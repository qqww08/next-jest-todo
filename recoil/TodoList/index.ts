import { atom } from "recoil";

export interface ITodo {
  id: number;
  name: string;
  checked: boolean;
}
const todoState = atom<ITodo[]>({
  key: "todoState",
  default: [],
});
export default todoState;
