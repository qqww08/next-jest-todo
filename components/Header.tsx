import styled from "styled-components";
import { useRecoilState } from "recoil";
import todoState, { ITodo } from "~/recoil/TodoList";
import { useCallback, useEffect, useState } from "react";

const Header = () => {
  const [value, setValue] = useState<string>("");
  const [checked, setChecked] = useState(false);
  const [todoValue, setTodoValue] = useRecoilState<ITodo[]>(todoState);

  const handleInputChange = useCallback((e) => {
    setValue(e.target.value);
  }, []);

  const handleAllCheckChange = useCallback(() => {
    setTodoValue((prev) => {
      return prev.map((item) => ({ ...item, checked: !checked }));
    });

    setChecked((prev) => !prev);
  }, [checked]);

  const handleInputKeyDown = useCallback(
    (e) => {
      if (e.keyCode === 13) {
        setValue("");
        setTodoValue((prev) => [...prev, { id: prev.length, name: value, checked: false }]);
      }
    },
    [value],
  );

  useEffect(() => {
    const todoFilter = todoValue.filter((item) => !item.checked);
    const isChecked = todoFilter.length > 0;
    if (isChecked) {
      setChecked(false);
    }
  }, [todoValue]);

  return (
    <__Header>
      <__Title>Todo List</__Title>
      <__InputWrapper>
        <__CheckBox type="checkbox" onChange={handleAllCheckChange} checked={checked} />
        <__Input placeholder="입력" onChange={handleInputChange} value={value} onKeyDown={handleInputKeyDown} />
      </__InputWrapper>
    </__Header>
  );
};

export default Header;
const __Header = styled.header`
  text-align: center;
`;
const __Title = styled.h1`
  font-size: 60px;
`;
const __InputWrapper = styled.div`
  display: flex;
  align-items: center;
`;
const __CheckBox = styled.input`
  width: 50px;
  height: 50px;
`;
const __Input = styled.input`
  width: 500px;
  height: 50px;
`;
