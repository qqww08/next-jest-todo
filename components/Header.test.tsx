import { fireEvent, render, screen } from "@testing-library/react";
import { RecoilRoot } from "recoil";
import Header from "./Header";

describe("Home", () => {
  test("render", () => {
    const { container } = render(
      <RecoilRoot>
        <Header />
      </RecoilRoot>,
    );

    expect(container).toMatchSnapshot();
  });
  test("input tag 에 입력한 값을 제대로 가지고 오고 있는지 체크", () => {
    render(
      <RecoilRoot>
        <Header />
      </RecoilRoot>,
    );

    const targetValue = "todo list";
    fireEvent.change(screen.getByTestId("test-input"), { target: { value: targetValue } });
    const input: HTMLInputElement = screen.getByDisplayValue(targetValue);
    expect(input.value).toEqual(targetValue);
  });

  test("엔터를 눌렀을때 recoil setter 가 정상적으로 작동하는지 test", () => {
    render(
      <RecoilRoot>
        <Header />
      </RecoilRoot>,
    );
  });
});
