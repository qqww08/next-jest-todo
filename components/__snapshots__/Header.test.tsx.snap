// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home render 1`] = `
<div>
  <header
    class="sc-2fa9c43c-0 fYHeri"
  >
    <h1
      class="sc-2fa9c43c-1 hiITkg"
    >
      Todo List
    </h1>
    <div
      class="sc-2fa9c43c-2 kmclcl"
    >
      <input
        class="sc-2fa9c43c-3 kOCMpp"
        data-testid="test-checkbox"
        type="checkbox"
      />
      <input
        class="sc-2fa9c43c-4 kMDneS"
        data-testid="test-input"
        placeholder="입력"
        value=""
      />
    </div>
  </header>
</div>
`;
